### DISCUSSION TITLE
Fuzzing: Vary arguments

### DISCUSSION THREAD

**Discussion by dloebl** [libvips/libvips#4095] [Ideas]
The recent improvements to the fuzzers made me think about the general way libvips is fuzzed by OSS-Fuzz.
It looks like that some parameters aren't really covered by the fuzzing targets. For example, the `page` or `n` argument are never passed to the `webpload` or `gifload`. This also applies to the saver (eg. a different `Q` for `jpegsave`).
```C
// option_string is empty
vips_image_new_from_buffer(data, size, "", nullptr)
// option list is empty
vips_image_write_to_buffer(image, SAVE_SUFFIX, &buf, &len, nullptr)
```
Ideally, the input data would be encoded in such a way that all arguments can be combined. I'm not really sure how to achieve this, since the inputs are essentially just images (at least as of now). For [cgif](https://github.com/dloebl/cgif) I made the API parameters part of the fuzzer input data (https://github.com/dloebl/cgif/pull/45). However, I'm not sure if that approach is viable for libvips.
Varying the arguments of loaders and savers should lead to better fuzzing coverage.

**Comment by jcupitt** [comment]
Hi @dloebl, yes, we do that for the mosaicing fuzzer, but you're right it's a long way from being comprehensive.

**Reply by kleisauke** [reply]
One consideration is that adding a feature similar to the mosaicing fuzzer might alter our existing seed corpus, potentially impacting the regression tests. IIRC, this was also a side effect of PR #3332, which led to certain OSS-Fuzz issues being closed and re-opened.

Another possible improvement for the fuzzers could be to support images with dimensions exceeding 100 pixels in width/height, see:
https://github.com/libvips/libvips/pull/3927#issuecomment-2043257488

**Reply by dloebl** [reply]
> One consideration is that adding a feature similar to the mosaicing fuzzer might alter our existing seed corpus

Makes sense!
I thought about adding an entirely new fuzzer to avoid this. I started working on it #4103

**Comment by kleisauke** [comment]
We do have some coverage for this through sharp, see for example:
https://github.com/google/oss-fuzz/blob/master/projects/sharp/fuzz.js

The only caveat is that fuzzing is done on the prebuilt binaries without any sanitizers. Nevertheless, it was able to catch GHSA-33qp-9pq7-9584.
